components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -18,91 +22,255 @@ import android.os.ParcelFileDescriptor
+                    DownloadNotification.createDownloadFailedNotification(context, download)
+                }
+
+                null -> { return }
+            }
         } catch (e: IOException) {
-            DownloadNotification.createDownloadFailedNotification(context, download.fileName)
+            DownloadNotification.createDownloadFailedNotification(context, download)
         }
+
         NotificationManagerCompat.from(context).notify(
+                downloadJobs[download.id]?.foregroundServiceId ?: 0,
+                notification
+        )
+
+        sendDownloadCompleteBroadcast(download.id)
+    }
+
+    private fun registerForUpdates() {
+        val filter = IntentFilter().apply {
+            addAction(ACTION_PAUSE)
+            addAction(ACTION_RESUME)
+            addAction(ACTION_CANCEL)
+            addAction(ACTION_TRY_AGAIN)
+            addAction(ACTION_OPEN)
+        }
+
+        context.registerReceiver(broadcastReceiver, filter)
+    }
+
+    private fun displayOngoingDownloadNotification(download: DownloadState) {
+        val ongoingDownloadNotification = DownloadNotification.createOngoingDownloadNotification(
             context,
-            COMPLETED_DOWNLOAD_NOTIFICATION_TAG,
-            notification
+            download
         )
 
-        val downloadID = intent.getLongExtra(EXTRA_DOWNLOAD_ID, -1)
-        sendDownloadCompleteBroadcast(downloadID)
+        // We want to startForeground so that the system is less likely to kill our service under memory pressure.
+        startForeground(
+            downloadJobs[download.id]?.foregroundServiceId ?: 0,
+            ongoingDownloadNotification
+        )
     }
 
-    private suspend fun performDownload(download: DownloadState) = withContext(IO) {
-        val headers = listOf(
-            CONTENT_TYPE to download.contentType,
-            CONTENT_LENGTH to download.contentLength?.toString(),
-            REFERRER to download.referrerUrl
-        ).mapNotNull { (name, value) ->
-            if (value.isNullOrBlank()) null else Header(name, value)
-        }.toMutableHeaders()
+    internal fun performDownload(download: DownloadState) {
+        val isResumingDownload = downloadJobs[download.id]?.currentBytesCopied ?: 0L > 0L
+        val headers = MutableHeaders()
 
-        val request = Request(download.url, headers = headers)
+        if (isResumingDownload) {
+            headers.append(RANGE, "bytes=${downloadJobs[download.id]?.currentBytesCopied}-")
+        }
 
+        val request = Request(download.url, headers = headers)
         val response = httpClient.fetch(request)
 
+        if (response.status != PARTIAL_CONTENT_STATUS && response.status != OK_STATUS) {
+            // We experienced a problem trying to fetch the file, send a failure notification
+            downloadJobs[download.id]?.currentBytesCopied = 0
+            downloadJobs[download.id]?.status = DownloadJobStatus.FAILED
+            return
+        }
```

Comment:
One thing I am worried about here is that a server responds with 200 OK but doesn't actually respect the byte range. I believe in your testing you mentioned this worked? Just not sure that's always the case. Should we double-check by making sure the response contains the `Content-Range` header?

Reply:
+1, I think we should check for `Content-Range` to ensure that range is observed.

Reply:
Added a check :)